File: threadring/src/threadring.c
/* Compile under Linux with
"cc threadring.sem.c -o threadring -lpthread -D Hops=100000 -D Nodes=1000"; run
with "./threadring". Does not run under macOS as unnamed semaphores
(using sem_init) are not supported.

Unnamed pthread semaphores are used for passing the token between node threads.
An array of split binary semaphores (i.e. at most one of them is 1) is used for
synchronization and for mutually exclusive access to global variable token with
the remaining number of hops. The node threads threads "do not know" about each
other, they only communicate via the semaphores and the token variable. When a
node thread decrements token to 0, it prints its id, otherwise it signals the
next node thread.

In C, a program terminates when the main program terminates. Here, the main
program waits for all node threads to terminate, which they do when they don't
expect any more tokens, i.e. the received token value is less than the number of
nodes.
*/

#include <pthread.h>
#include <stdio.h>
#include <semaphore.h>
#include <stdlib.h>

sem_t *node_sem;
int token;
int Nodes;
void *node(void *argptr) {
  long nid = (long) argptr;
  do {
    sem_wait(&node_sem[nid]);
    // printf("%d received %d\n", nid, token);
    if (token > 0) {
      token -= 1;
      sem_post(&node_sem[(nid + 1) % Nodes]);
    } else printf("%ld\n", nid);
  } while (token >= Nodes - 1);
}
int main(int argc, char *argv[]) {
  if(argc < 3){
      printf("The Usage: %s Hops Nodes\n", argv[0]);
      exit(1);
  }
  token = atoi(argv[1]);
  Nodes = atoi(argv[2]);
  node_sem = (sem_t *)malloc(sizeof(sem_t)*Nodes);
  pthread_t tid[Nodes];
  for (long n = 0; n < Nodes; n++) {
    sem_init(&node_sem[n], 0, 0);
    pthread_create(&tid[n], NULL, node, (void *) n);
  }
  sem_post(&node_sem[0]);
  for (long n = 0; n < Nodes; n++) {
    pthread_join(tid[n], NULL);
  }
}

File: threadring/src/threadring.java
/* Compile under Linux and macOS with "javac threadring.java"; run with
"java ThreadRing"

Java monitors with signalling by notify and wait are used. Class Node defines
active objects in the sense that they have a method that can be called and each
has a thread that calls another object. The thread and the method of each object
are synchronized, i.e. run mutually exclusive. Each object passes the token by
calling the pass method of the next object. That method notifies the object's
thread to pass on the token to its next object. A token value of -1 is used by
the thread to check if a valid token has already been passed or the thread needs
to wait for it.

In Java, a program terminates when all threads terminate. Each node thread
terminates when it won't receive any more tokens, which is when the received
token is less than the number of nodes.
*/
class Node extends Thread {
  Node next;
  int token = - 1;
  int n;
  int nodes; 
  Node(int n, int nn) {
    this.n = n; this.nodes = nn; start();
  }
  synchronized void pass(int t) {
    token = t; notify();
  }
  synchronized public void run() {
    if (token == - 1)
      try {wait();} catch (InterruptedException e) {}
    while (true) {
      // System.out.println(n + " received " + token);
      if (token > 0) next.pass(token - 1);
      else System.out.println(n);
      if (token < this.nodes) break;
      try {wait();} catch (InterruptedException e) {}
    }
  }
}
class ThreadRing {
  // static final int Hops = 100000;
  // static final int Nodes = 1000;
  public static void main(String args[]) {
    if(args.length < 2){
        System.err.println("Usage: java ThreadRing Hops Nodes\n");
        return;
    }
    int Hops = Integer.parseInt(args[0]);
    int Nodes = Integer.parseInt(args[1]);
    Node nd0 = new Node(0, Nodes);
    Node nd = nd0;
    for (int n = 1; n < Nodes; n++) {
      nd.next = new Node(n, Nodes); nd = nd.next;
    }
    nd.next = nd0;
    nd0.pass(Hops);
  }
}
